Meadowlark-js/packages/meadowlark-core/src/handler/Upsert.ts

```diff
@@ -3,105 +3,115 @@
-      return errorResponse as APIGatewayProxyResult;
+      writeDebugStatusToLog(moduleName, frontendRequest, 'create', errorResponse.statusCode, JSON.stringify(jwtStatus));
+      return errorResponse as FrontendResponse;
     }
 
-    const pathComponents: PathComponents | null = pathComponentsFrom(event.path);
+    const pathComponents: PathComponents | null = pathComponentsFrom(frontendRequest.path);
 
     if (pathComponents === null) {
-      writeDebugStatusToLog(moduleName, context, 'create', 404);
+      writeDebugStatusToLog(moduleName, frontendRequest, 'create', 404);
       return { body: '', statusCode: 404 };
     }
 
-    if (event.body == null) {
+    if (frontendRequest.body == null) {
       const message = 'Missing body';
-      writeDebugStatusToLog(moduleName, context, 'create', 400, message);
+      writeDebugStatusToLog(moduleName, frontendRequest, 'create', 400, message);
       return { body: JSON.stringify({ message }), statusCode: 400 };
     }
 
     let body: object = {};
     try {
-      body = JSON.parse(event.body);
+      body = JSON.parse(frontendRequest.body);
     } catch (error) {
       const message = 'Malformed body';
-      writeDebugStatusToLog(moduleName, context, 'create', 400, message);
+      writeDebugStatusToLog(moduleName, frontendRequest, 'create', 400, message);
       return { body: JSON.stringify({ message }), statusCode: 400 };
     }
 
     const { documentInfo, errorBody, headerMetadata } = await validateResource(pathComponents, body);
     if (errorBody != null) {
       const statusCode = documentInfo === NoDocumentInfo ? 404 : 400;
-      writeDebugStatusToLog(moduleName, context, 'create', statusCode, errorBody);
+      writeDebugStatusToLog(moduleName, frontendRequest, 'create', statusCode, errorBody);
       return { body: errorBody, statusCode, headers: headerMetadata };
     }
 
     const resourceId = documentIdForDocumentInfo(documentInfo);
-    const { result, failureMessage } = await getBackendPlugin().upsertDocument({
+    const request: UpsertRequest = {
       id: resourceId,
       documentInfo,
       edfiDoc: body,
-      validate: event.headers['reference-validation'] !== 'false',
+      validate: frontendRequest.headers['reference-validation'] !== 'false',
       security: {
         ...newSecurity(),
         isOwnershipEnabled: jwtStatus.isOwnershipEnabled,
         clientName: jwtStatus.subject,
       },
-      traceId: context.awsRequestId,
-    });
-    if (result === 'INSERT_SUCCESS') {
-      writeDebugStatusToLog(moduleName, context, 'create', 201);
-      const location = `${event.path}${event.path.endsWith('/') ? '' : '/'}${resourceId}`;
+      traceId: frontendRequest.traceId,
+    };
+
+    beforeUpsertDocument(request);
+    const result: UpsertResult = await getDocumentStore().upsertDocument(request);
+    afterUpsertDocument(request, result);
```

Comment:
If change data capture becomes a hassle in some way, we could also use this notification process to log to Kafka.

Reply:
👍